apps/src/lab2/progress/ProgressManager.ts

```diff
@@ -86,7 +86,21 @@ export default class ProgressManager {
   }
 
   updateProgress(): void {
-    if (!this.currentValidations || !this.validator) {
+    if (!this.validator) {
+      return;
+    }
+    const exemplarSettings = this.exemplarSettings;
+    const shouldValidateExemplar =
+      !getAppOptionsEditingExemplar() && exemplarSettings?.validationEnabled;
+
+    // If there are no validations, we still might run the exemplar validation.
+    if (!this.currentValidations?.length && shouldValidateExemplar) {
+      this.setValidationStateWithExemplar();
```

Comment:
nit: should we store `didPassExemplarValidation` and the appropriate exemplar validation message here so we don't have to compute it twice? Something like:

```
let passedExemplar = false;
let exemplarMessage = '';
if (shouldValidateExemplar) {
  passedExemplar = this.validator.didPassExemplarValidation();
  exemplarMessage = passedExemplar ? ... : ...;
}

if (!this.currentValidations) {
  // Set exemplar validation status if necessary and return early
  if (shouldValidateExemplar) {
    this.currentValidationState = { ...this.currentValidationState, passed: passedExemplar, message: ... };
    this.onProgressChange()
  }
  return;
}
...
// Use already computed values of passedExemplar and exemplarMessage while looping through conditions
```